### PR TITLE
Use Worker instead of ZkWorker whenever possible

### PR DESCRIPTION
**Pull request by drcrallen**
~~Depends on https://github.com/druid-io/druid/pull/1953~~

### CODE REVIEW

File: services/src/main/java/io/druid/cli/ServerRunnable.java [hunk 1]
```diff
@@ -45,5 +45,8 @@ public void run()
     catch (Exception e) {
       throw Throwables.propagate(e);
     }
+    finally {
+      lifecycle.stop();
```
Comment (drcrallen) [inline services/src/main/java/io/druid/cli/ServerRunnable.java]: I think this line snuck in during some of my commit shuffling. I'll split it out before final rebase

File: services/src/main/java/io/druid/cli/CliPeon.java [hunk 2]
```diff
@@ -121,39 +120,13 @@ public void configure(Binder binder)
...
-                                 .to(NoopChatHandlerProvider.class).in(LazySingleton.class);
-            binder.bind(ServiceAnnouncingChatHandlerProvider.class).in(LazySingleton.class);
-
-            binder.bind(NoopChatHandlerProvider.class).in(LazySingleton.class);
-
-            binder.bind(TaskToolboxFactory.class).in(LazySingleton.class);
-
-            JsonConfigProvider.bind(binder, "druid.indexer.task", TaskConfig.class);
-            JsonConfigProvider.bind(binder, "druid.peon.taskActionClient.retry", RetryPolicyConfig.class);
+            bindTaskChatter(binder);
```
Comment (drcrallen) [inline services/src/main/java/io/druid/cli/CliPeon.java]: Will revert this one. Stale changes from old commit

File: indexing-service/src/main/java/io/druid/indexing/worker/http/WorkerResource.java [hunk 3]
```diff
@@ -54,18 +57,60 @@
...
+        worker.getCapacity(),
+        DISABLED_VERSION,
+        enabledWorker.getLastCompletedTaskTime()
+    )
+    {
+      @Override
+      @JsonProperty
+      public String getHost()
+      {
+        return enabledWorker.getHost();
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/worker/http/WorkerResource.java]: Why does this have to happen? These properties seems like they should work on their own for everything except lastCompletedTaskTime.

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/worker/http/WorkerResource.java]: Also in light of the need for overriding it'd probably be cleaner to generate the `disabledWorker` every time someone posts a disable call. That happens so infrequently that there is not really much need to optimize it…

File: indexing-service/src/main/java/io/druid/indexing/worker/http/WorkerResource.java [hunk 4]
```diff
@@ -164,18 +209,21 @@ public Response doGetLog(
       @QueryParam("offset") @DefaultValue("0") long offset
   )
   {
-    final Optional<ByteSource> stream = taskRunner.streamTaskLog(taskid, offset);
+    if (!(taskRunner instanceof TaskLogStreamer)) {
+      return Response.status(501).entity("Log streaming not supported").build();
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/worker/http/WorkerResource.java]: Include the TaskRunner class name, perhaps

File: server/src/main/java/io/druid/server/initialization/ZkPathsConfig.java [hunk 5]
```diff
@@ -90,7 +90,7 @@ public String getConnectorPath()
     return (null == connectorPath) ?  defaultPath("connector") : connectorPath;
   }
 
-  protected String defaultPath(final String subPath)
+  public String defaultPath(final String subPath)
```
Comment (gianm) [inline server/src/main/java/io/druid/server/initialization/ZkPathsConfig.java]: what was this for?

File: server/src/main/java/io/druid/server/initialization/jetty/ChatHandlerServerModule.java [hunk 6]
```diff
@@ -45,7 +45,7 @@
   private static final Logger log = new Logger(ChatHandlerServerModule.class);
 
   @Inject
-  private Properties properties;
+  private Properties properties = new Properties(System.getProperties());
```
Comment (gianm) [inline server/src/main/java/io/druid/server/initialization/jetty/ChatHandlerServerModule.java]: what's this change for? It seems like it would break property injection…

File: services/src/main/java/io/druid/cli/CliOverlord.java [hunk 7]
```diff
@@ -104,7 +105,7 @@ public CliOverlord()
   }
 
   @Override
-  protected List<? extends Module> getModules()
+  public List<? extends Module> getModules()
```
Comment (gianm) [inline services/src/main/java/io/druid/cli/CliOverlord.java]: what's this for?

**Comment by gianm** [comment]
Broadly looks good, I assume this is needed for #2246. would like to understand the property injection change better though.

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java [hunk 8]
```diff
@@ -91,6 +92,13 @@ public void addTerminateEvent(AutoScalingData data)
     }
   }
 
+  public void addAllEvents(Collection<ScalingEvent> events)
```
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java]: why is this needed?

File: indexing-service/src/main/java/io/druid/indexing/worker/http/WorkerResource.java (hunk 3 above)
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/worker/http/WorkerResource.java]: that's reasonable

File: indexing-service/src/main/java/io/druid/indexing/worker/http/WorkerResource.java (hunk 4 above)
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/worker/http/WorkerResource.java]: reasonable

File: server/src/main/java/io/druid/server/initialization/ZkPathsConfig.java (hunk 5 above)
Comment (drcrallen) [inline server/src/main/java/io/druid/server/initialization/ZkPathsConfig.java]: accidentally included when I migrated PR stuff. will fix.

In general it is nice for extensions to be able to build a default ZK path using zkPathsConfig without extending zkpathsconfig itself, I will remove since it is out of scope of this pr

File: server/src/main/java/io/druid/server/initialization/jetty/ChatHandlerServerModule.java (hunk 6 above)
Comment (drcrallen) [inline server/src/main/java/io/druid/server/initialization/jetty/ChatHandlerServerModule.java]: will double-check, I do not remember right now

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java (hunk 8 above)
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ScalingStats.java]: removed

**Comment by drcrallen** [comment]
@gianm  / @himanshug rebased.

**Comment by drcrallen** [comment]
failedin ```Tests run: 7, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.047 sec <<< FAILURE! - in io.druid.indexing.overlord.config.RemoteTaskRunnerConfigTest
testHashCode(io.druid.indexing.overlord.config.RemoteTaskRunnerConfigTest)  Time elapsed: 0.013 sec  <<< FAILURE!
java.lang.AssertionError: Values should be different. Actual: 1709774762
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failEquals(Assert.java:185)
	at org.junit.Assert.assertNotEquals(Assert.java:161)
	at org.junit.Assert.assertNotEquals(Assert.java:198)
	at org.junit.Assert.assertNotEquals(Assert.java:209)
	at io.druid.indexing.overlord.config.RemoteTaskRunnerConfigTest.testHashCode(RemoteTaskRunnerConfigTest.java:335)
```

investigating

**Comment by drcrallen** [comment]
was error in new code. fixed

File: services/src/main/java/io/druid/cli/CliOverlord.java (hunk 7 above)
Comment (drcrallen) [inline services/src/main/java/io/druid/cli/CliOverlord.java]: I think some other PR I was working on , removing

**Comment by drcrallen** [comment]
@himanshug  / @gianm  updated

File: common/src/main/java/io/druid/concurrent/Execs.java [hunk 9]
```diff
@@ -86,6 +86,18 @@ public static ThreadFactory makeThreadFactory(@NotNull String nameFormat, @Nulla
     return builder.build();
   }
 
+  public static AutoCloseable asAutoCloseableShutdownNow(final ExecutorService service)
```
Comment (gianm) [inline common/src/main/java/io/druid/concurrent/Execs.java]: appears unused

File: indexing-service/src/main/java/io/druid/indexing/overlord/ForkingTaskRunnerFactory.java [hunk 10]
```diff
@@ -34,6 +34,7 @@
  */
 public class ForkingTaskRunnerFactory implements TaskRunnerFactory<ForkingTaskRunner>
 {
+  public static final String TYPE_NAME = "local";
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/ForkingTaskRunnerFactory.java]: this is a bit odd imo, it's only a "local" runner on the overlord. On the middleManager it's the normal runner.

Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/ForkingTaskRunnerFactory.java]: so I think whether it's "local" or whatever properly belongs to the Overlord/MM guice modules and not to the runner factory itself.

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java [hunk 11]
```diff
@@ -1066,9 +1076,25 @@ private void taskComplete(
...
 
-  public List<ZkWorker> getLazyWorkers()
+  @Override
+  public Collection<Worker> getLazyWorkers()
   {
-    return ImmutableList.copyOf(lazyWorkers.values());
+    return ImmutableList.copyOf(getWorkerFromZK(lazyWorkers.values()));
+  }
+
+  protected Collection<Worker> getWorkerFromZK(Collection<ZkWorker> workers)
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: could this be `private`? RTR doesn't have subclasses as far as I know

File: indexing-service/src/main/java/io/druid/indexing/overlord/ForkingTaskRunnerFactory.java (hunk 10 above)
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/ForkingTaskRunnerFactory.java]: reasonable, removed

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java (hunk 11 above)
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: I'm going to go the other way here actually, and make this a public static, even if it isn't used elsewhere

**Comment by drcrallen** [comment]
@gianm updated / addressed

**Comment by gianm** [comment]
:+1:

**Comment by drcrallen** [comment]
test failed from updated unit tests with missing parameters in master. fixed

File: indexing-service/src/main/java/io/druid/indexing/overlord/ZkWorker.java [hunk 12]
```diff
@@ -129,7 +129,7 @@ public int getCurrCapacityUsed()
   @JsonProperty
   public DateTime getLastCompletedTaskTime()
   {
-    return lastCompletedTaskTime.get();
+    return worker.get().getLastCompletedTaskTime();
```
Comment (nishantmonu51) [inline indexing-service/src/main/java/io/druid/indexing/overlord/ZkWorker.java]: can't see the variable lastCompletedTaskTime getting deleted ?

**Comment by nishantmonu51** [comment]
:+1:, only one minor nit about unused variable.

File: indexing-service/src/main/java/io/druid/indexing/overlord/ZkWorker.java (hunk 12 above)
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/ZkWorker.java]: Class field was leftover value that was accidentally used in toString, removed

File: indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java [hunk 13]
```diff
@@ -325,7 +327,13 @@ public void stop()
     return ImmutableList.of();
   }
 
-  public Collection<ZkWorker> getWorkers()
+  @Override
+  public Collection<Worker> getWorkers()
```
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: @drcrallen this change made the overlord `/druid/indexer/v1/workers` API no longer return a `runningTasks` key for each worker, which may confuse some folks' rolling update scripts and definitely makes the web UI less useful.

how do you feel about having the OverlordResource specifically call `getZkWorkers` if the runner is an RTR? or would it make more sense to add a list of running tasks to the Worker object?

Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java]: ((spoke in chat)) since io.druid.indexing.overlord.http.OverlordResource#getWorkers already special-cases the RTR, suggest making it get zk workers in that case until a more complete solution can be put in place.
